lib/Sema/CSDiagnostics.cpp

```diff
@@ -101,6 +101,35 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
   return nullptr;
 }
 
+// TODO: Replace duplications of this logic with calls to this.
+Optional<SelectedOverload> FailureDiagnostic::getChoiceFor(Expr *expr) {
+  auto &cs = getConstraintSystem();
+  ConstraintLocator *locator = nullptr;
+
+  if (auto *call = dyn_cast<CallExpr>(expr)) {
+    auto *fnExpr = call->getFn();
+    return getChoiceFor(fnExpr);
+  } else if (auto *UDE = dyn_cast<UnresolvedDotExpr>(expr)) {
+    locator = cs.getConstraintLocator(UDE, ConstraintLocator::Member);
+  } else if (auto *UME = dyn_cast<UnresolvedMemberExpr>(expr)) {
+    locator = cs.getConstraintLocator(UME, ConstraintLocator::UnresolvedMember);
+  } else if (auto *TE = dyn_cast<TypeExpr>(expr)) {
+    locator = cs.getConstraintLocator(call,
```

Comment:
There is no `call` in this scope, so it wouldn't compile, right? Did you intend to handle `TypeExpr` in the `CallExpr` branch? That would be a good idea. As a related note - use `ApplyExpr` there instead of `CallExpr` that is going to cover operator use like `1 + 1` where `+` is going to be `OverloadedDeclRefExpr`.

Reply:
Ahh, this does compile because call is still bound here, it just won't work because it'll never be reached in the case where the call happens. I'm amazed this didn't break any tests, nice spot.